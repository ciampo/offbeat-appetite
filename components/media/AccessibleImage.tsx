import React, { memo } from 'react';
import { urlFor } from '../../sanity/get-image-url';

import { AccessibleImageResponsiveConfig } from './image-responsive-configurations';
import { SanityAccessibleImage } from '../../typings';

function generateUrl({
  image,
  width,
  forceRatio,
}: {
  image: SanityAccessibleImage;
  width: number;
  forceRatio?: number;
}): string {
  let builder = urlFor(image)
    // Serve WebP when possible
    .auto('format')
    // When specifying both width/height, crops the image to fit
    .fit('crop');

  if (forceRatio) {
    builder = builder.size(Math.round(width), Math.round(width / forceRatio));
  } else {
    builder = builder.width(Math.round(width));
  }

  return builder.url() as string;
}

type AccessibleImageProps = {
  image: SanityAccessibleImage;
  responsiveConfig: AccessibleImageResponsiveConfig;
  darker?: boolean;
  lazy?: boolean;
  className?: string;
  style?: React.CSSProperties;
  [key: string]: unknown;
};

const AccessibleImage: React.FC<AccessibleImageProps> = ({
  image,
  lazy,
  darker,
  responsiveConfig: { exports, sizes, forceRatio },
  style,
  className = 'relative w-full h-0',
  ...props
}) => {
  if (!exports.length || !sizes.length) {
    console.warn(
      `AccessibleImage needs valid options to be rendered [raw url: ${image.asset.url}]`
    );
    return null;
  }

  if (sizes[sizes.length - 1].queryMinWidth) {
    console.warn(
      'The last element of the sizes config should not have a `queryMinWidth` attribute'
    );
    return null;
  }

  return (
    <div
      {...props}
      className={['contain-l-p', className].filter(Boolean).join(' ')}
      style={{
        // Fills the space when the image is lazily loaded
        backgroundColor: image.asset.metadata.palette.dominant.background,
        // Force aspect ratio (avoids layout shifting)
        paddingBottom: `${
          Math.round(100000 / (forceRatio || image.asset.metadata.dimensions.aspectRatio)) / 1000
        }%`,
        ...style,
      }}
      data-testid="image-wrapper"
    >
      <img
        className="absolute top-0 left-0 w-full h-full object-cover"
        style={{
          objectPosition: image.hotspot
            ? `${image.hotspot.x * 100}% ${image.hotspot.y * 100}%`
            : '50% 50%',
        }}
        alt={image.alt}
        // Fallback src when srcset is not supported
        src={generateUrl({
          image,
          width: exports[0],
          forceRatio,
        })}
        srcSet={exports
          .map(
            (width) =>
              `${generateUrl({
                image,
                width,
                forceRatio,
              })} ${width}w`
          )
          .join(', ')}
        sizes={sizes
          .map(
            ({ width, queryMinWidth }) =>
              `${queryMinWidth ? `(min-width: ${queryMinWidth}) ` : ''}${width}`
          )
          .join(',')}
        // Lazy-loading related attributes
        loading={lazy ? 'lazy' : undefined}
        width={image.asset.metadata.dimensions.width}
        height={
          forceRatio
            ? Math.round(image.asset.metadata.dimensions.width / forceRatio)
            : image.asset.metadata.dimensions.height
        }
      />
      {darker && (
        <div
          className="absolute inset-0 bg-black bg-opacity-35 pointer-events-none"
          aria-hidden="true"
          data-testid="image-scrim"
        ></div>
      )}
    </div>
  );
};

const MemoizedAccessibleImage = memo(AccessibleImage);
export default MemoizedAccessibleImage;
