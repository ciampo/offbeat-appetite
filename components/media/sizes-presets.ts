import { sharedTheme as twTheme } from '../../tailwind.shared';

function remToPx(rem: string): number {
  return parseFloat(rem) * 16;
}

function pxToRem(px: string): number {
  return parseFloat(px) / 16;
}

const sectionPaddingH = twTheme.spacing['6'];
const sectionPaddingHPx = remToPx(sectionPaddingH);
// const maxWidthXs = twTheme.maxWidth.xs;
// const maxWidthXsPx = remToPx(maxWidthXs);

export type AccessibleImageSizeConfig = {
  queryMinWidth?: string;
  width: string;
};

export type AccessibleImageResponsiveConfig = {
  exports: number[];
  sizes: AccessibleImageSizeConfig[];
  forceRatio?: number;
};

function sortBreakpoints({
  breakpoints,
  sorting = 'asc',
  unit,
  addDoubleRes,
}: {
  breakpoints: string[];
  sorting?: 'asc' | 'desc';
  unit?: string;
  addDoubleRes?: boolean;
}): (string | number)[] {
  let numbers = breakpoints.map((s: string) => parseInt(s, 10));

  if (addDoubleRes) {
    numbers = Array.from(
      new Set(numbers.reduce((arr, bp: number) => [...arr, bp, bp * 2], [] as number[]))
    );
  }

  return (
    numbers
      // sort asc/desc
      .sort((x, y) => (sorting === 'asc' ? (x > y ? 1 : -1) : x < y ? 1 : -1))
      // potentially add the unit
      .map((n) => (unit ? `${n}${unit}` : n))
  );
}

// =============================================================================
// PRESETS
// =============================================================================

// Used in full bleed images (e.g. page hero)
export const fullBleedImageResponsiveConfig: AccessibleImageResponsiveConfig = {
  exports: sortBreakpoints({
    breakpoints: [
      '320',
      twTheme.screens.xsm,
      twTheme.screens.sm,
      twTheme.screens.md,
      twTheme.screens.lg,
      twTheme.screens.xl,
    ],
    addDoubleRes: true,
  }) as number[],
  sizes: [{ width: '100vw' }],
  forceRatio: 16 / 9,
};

// Used in blog post preview tiles
export const blogPostPreviewResponsiveConfig: AccessibleImageResponsiveConfig = {
  exports: sortBreakpoints({ breakpoints: ['320'] }) as number[],
  sizes: [{ width: `${pxToRem('320')}rem` }],
  forceRatio: 3 / 2,
};

// Used in blog post content for 100% wide images
export const contentFullWidthResponsiveConfig: AccessibleImageResponsiveConfig = {
  exports: sortBreakpoints({
    breakpoints: [
      `${375 - 2 * sectionPaddingHPx}`,
      `${remToPx(twTheme.maxWidth['2xl']) - 2 * sectionPaddingHPx}`,
      `${remToPx(twTheme.maxWidth['3xl']) - 2 * sectionPaddingHPx}`,
    ],
    addDoubleRes: true,
  }) as number[],
  sizes: [
    {
      queryMinWidth: twTheme.maxWidth['screen-lg'],
      width: `calc(${twTheme.maxWidth['3xl']} - 2 * ${sectionPaddingH})`,
    },
    {
      queryMinWidth: `${remToPx(twTheme.maxWidth['2xl'])}px`,
      width: `calc(${twTheme.maxWidth['2xl']} - 2 * ${sectionPaddingH})`,
    },
    {
      width: `calc(100vw - 2 * ${sectionPaddingH})`,
    },
  ],
};
