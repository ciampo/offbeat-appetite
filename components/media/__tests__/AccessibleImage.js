import React from 'react';
import { render } from '@testing-library/react';
import AccessibleImage from '../AccessibleImage';
import { contentFullWidthResponsiveConfig } from '../sizes-presets';

const testImage = {
  _type: 'image',
  alt: 'The best koala in town ',
  asset: {
    _id: 'image-673168dce625d591c39090208482ae76210f3b2f-1280x720-jpg',
    assetId: '673168dce625d591c39090208482ae76210f3b2f',
    extension: 'jpg',
    metadata: {
      dimensions: {
        aspectRatio: 1.7777777777777777,
        height: 720,
        width: 1280,
      },
      palette: {
        dominant: {
          background: '#c3b3a7',
          foreground: '#000',
        },
      },
    },
    url:
      'https://cdn.sanity.io/images/gpe6axmm/production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg',
  },
  crop: {
    _type: 'sanity.imageCrop',
    bottom: 0,
    left: 0,
    right: 0,
    top: 0,
  },
  hotspot: {
    _type: 'sanity.imageHotspot',
    height: 0.3345485056274047,
    width: 0.1666666666666658,
    x: 0.6260162601625998,
    y: 0.2866215316957439,
  },
};

test('renders', () => {
  const { container, getByRole } = render(
    <AccessibleImage image={testImage} responsiveConfig={contentFullWidthResponsiveConfig} />
  );

  expect(getByRole('img')).toHaveAttribute('alt', testImage.alt);

  expect(container).toMatchSnapshot();
});
