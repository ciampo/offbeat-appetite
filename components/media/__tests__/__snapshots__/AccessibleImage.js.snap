// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="relative w-full h-0"
    style="background-color: rgb(195, 179, 167); padding-bottom: 56.25%;"
  >
    <img
      alt="The best koala in town "
      class="absolute top-0 left-0 w-full h-full object-cover"
      height="720"
      sizes="(min-width: 1024px) calc(48rem - 2 * 1.5rem),(min-width: 672px) calc(42rem - 2 * 1.5rem),calc(100vw - 2 * 1.5rem)"
      src="https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=327&fit=crop&auto=format"
      srcset="https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=327&fit=crop&auto=format 327w, https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=624&fit=crop&auto=format 624w, https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=654&fit=crop&auto=format 654w, https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=720&fit=crop&auto=format 720w, https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=1248&fit=crop&auto=format 1248w, https://cdn.sanity.io/images//production/673168dce625d591c39090208482ae76210f3b2f-1280x720.jpg?w=1440&fit=crop&auto=format 1440w"
      style="object-position: 62.601626016259985% 28.66215316957439%;"
      width="1280"
    />
  </div>
</div>
`;
